storage.go

```diff
@@ -3,34 +3,297 @@ package minio
+		fileVersion = opt.FileVersion
+	}
+	readObject, err := s.client.GetObject(ctx, s.bucketName, objectPath, minio.GetObjectOptions{VersionID: fileVersion})
+	if err != nil {
+		return n, err
+	}
+	defer readObject.Close()
+
+	if opt.HasOffset {
+		_, err = readObject.Seek(opt.Offset, 0)
+		if err != nil {
+			return n, err
+		}
+	}
+
+	rc = readObject
+
+	if opt.HasSize {
+		rc = iowrap.LimitReadCloser(rc, opt.Size)
+	}
+
+	if opt.HasIoCallback {
+		rc = iowrap.CallbackReadCloser(rc, opt.IoCallback)
+	}
+
+	return io.Copy(w, readObject)
 }
 
 func (s *Storage) stat(ctx context.Context, path string, opt pairStorageStat) (o *Object, err error) {
-	panic("not implemented")
+	defer func() {
+		if err != nil {
+			err = s.formatError("stat", err, path)
+		}
+	}()
+
+	var fileVersion string
+	if opt.HasFileVersion {
+		fileVersion = opt.FileVersion
+	}
+
+	objectType := "file"
+	if opt.HasObjectType {
+		objectType = opt.ObjectType
+	}
+
+	uniquePath := s.getAbsPath(path)
+	dir := filepath.ToSlash(path)
+	switch objectType {
+	case "file":
+		objInfo, err := s.client.StatObject(ctx, s.bucketName, uniquePath, minio.StatObjectOptions{VersionID: fileVersion})
+		if err != nil {
+			return o, err
+		}
+		o = s.newObject(true)
+		o.ID = uniquePath
+		o.Path = path
+		o.Mode |= ModeRead
+		o.SetEtag(objInfo.ETag)
+		o.SetLastModified(objInfo.LastModified)
+		o.SetContentLength(objInfo.Size)
+	case "directory":
+		// minio couldn't stat/delete dir object directly
+		// ref: https://github.com/minio/minio-go/issues/803
+		// using list to found
+
+		// list will return key like prefix/dir/
+		// with '/' suffix and without '/' prefix
+		// so handle refer object name first
+
+		objKey := uniquePath
+		if strings.HasPrefix(objKey, "/") {
+			objKey = objKey[1:]
+		}
+		if !strings.HasSuffix(objKey, "/") {
+			objKey = objKey + "/"
+		}
+
+		ctxWithCancel, cancel := context.WithCancel(ctx)
+		defer func() {
+			// ctxWithCancel will return none-nil if closed/cancelled ...
+			// return nil if still opening
+			if ctxWithCancel.Err() == nil {
+				cancel()
+			}
+		}()
+
+		findChan := s.client.ListObjects(ctx, s.bucketName, minio.ListObjectsOptions{
+			Prefix:    uniquePath,
+			Recursive: false,
+		})
+
+		for object := range findChan {
+			if object.Err != nil {
+				cancel()
+				return
+			}
+
+			// return key like prefix/dir/ whit '/' suffix and without '/' prefix
+			if object.Key == objKey {
+				// already found, cancel
+				cancel()
+				o = s.formatFolderObject(&object, uniquePath, dir)
+				break
+			}
+		}
+	default:
+		return o, services.PairUnsupportedError{Pair: Pair{Key: "object_type", Value: opt.ObjectType}}
+	}
+
+	return
 }
 
 func (s *Storage) write(ctx context.Context, path string, r io.Reader, size int64, opt pairStorageWrite) (n int64, err error) {
-	panic("not implemented")
+	rp := s.getAbsPath(path)
+
+	uploadInfo, err := s.client.PutObject(ctx, s.bucketName, rp, r, size, minio.PutObjectOptions{ContentType: "application/octet-stream"})
```

Comment:
Please support `content-type` pair instead of hardcode to `application/octet-stream`.